man/visage-app.5
.\" generated with Ronn/v0.7.3
.\" http://github.com/rtomayko/ronn/tree/0.7.3
.
.TH "VISAGE\-APP" "5" "July 2013" "" ""
.
.SH "NAME"
\fBvisage\-app\fR \- Run a standalone instance of Visage
.
.SH "SYNOPSIS"
\fBvisage\-app\fR \fIstart|genapache\fR
.
.SH "DESCRIPTION"
Visage is a web interface for viewing collectd statistics\.
.
.P
It also provides a JSON interface onto collectd\'s RRD data, giving you an easy way to mash up the data\.
.
.SH "OPTIONS"
.
.TP
\fBstart\fR
Start visage on \fB127\.0\.0\.1:9292\fR
.
.TP
\fBgenapache\fR
Generate a vhost suitable for Passenger + Apache\. Will load Visage from the installed gem\.
.
.TP
\fBhelp\fR
Display this man page\.
.
.SH "ENVIRONMENT"
.
.TP
\fBCONFIG_PATH\fR
Where to look for configuration files\. Added to the beginning of config load path, so files found in here take precedence over shipped config files\.
.
.TP
\fBRRDDIR\fR
Where to look for collectd\'s RRDs\. Must correspond with collectd\'s \fBrrdtool\fR plugin\'s \fBDataDir\fR value\.
.
.TP
\fBCOLLECTDSOCK\fR
Where to look for collectd\'s Unix socket\. Used by Visage to issue \fBFLUSH\fR command to collectd to present more up to date statistics\.
.
.TP
\fBRRDCACHEDSOCK\fR
Where to look for rrdcached\'s Unix socket\. Used by Visage to issue \fBFLUSH\fR command to rrdcached to present more up to date statistics\.
.
.TP
\fBVISAGE_DATA_BACKEND\fR
Specify which data backend Visage should use to retrieve statistics\. An up to date list of backends can be found by looking in \fBlib/visage\-app/data\fR\. At the time of writing, the shipped backends are RRD and Mock\.
.
.SH "AUTHOR"
Lindsay Holmwood \fIlindsay@holmwood\.id\.au\fR
.
.SH "COPYRIGHT"
MIT Licensed\.


CHANGELOG.md
2.1.0 - 2012/04/20
 - Add support for drawing 95th percentile lines on graphs
   Thanks to Jesse Reynolds for the feature addition!

2.0.5 - 2012/03/30
 - Fix a bug where metrics occasionally won't appear in the builder interface.
   Thanks to Jesse Reynolds for the patch!
 - Add documentation on how to do a release, for when I get hit by a bus.

2.0.4 - 2012/03/27
 - Add extra input validation in the builder, so invalid data can't be entered
   into `profiles.yaml`, per [issue 92](https://github.com/auxesis/visage/issues/92)

2.0.2 - 2012/03/26
 - Builder bugfix for [issue 92](https://github.com/auxesis/visage/issues/92),
   that stopped the builder from presenting the save profile dialog.

2.0.0 - 2012/03/25
 - New builder interface, optimised for metric discovery + investigatory work
 - Backwards incompatible profile.yaml file format change
 - Chart performance optimisations
 - Visual improvements to chart layout + design
 - Bump versions of dependencies
 - Improve the installation documentation for non-Ubuntu platforms

1.0.0 - 2011/06/11

 - 1.0.0 release! Production ready.



lib/visage-app.rb
#!/usr/bin/env ruby

require 'pathname'
@root = Pathname.new(File.dirname(__FILE__)).expand_path
$: << @root.to_s

require 'sinatra/base'
require 'sinatra/reloader'
require 'haml'
require 'yajl/json_gem'
require 'visage-app/models/profile'
require 'visage-app/helpers'
require 'visage-app/config'
require 'visage-app/data'
require 'visage-app/upgrade'

module Visage
  class Application < Sinatra::Base
    @root = Pathname.new(File.dirname(__FILE__)).expand_path
    set :public_folder, @root.join('visage-app/public')
    set :views,         @root.join('visage-app/views')

    enable :logging

    helpers Sinatra::LinkToHelper
    helpers Sinatra::PageTitleHelper
    helpers Sinatra::RequireJSHelper
    helpers Sinatra::RequireCSSHelper
    helpers Sinatra::FormatHelper

    configure do
      Visage::Config.use do |c|
        # FIXME: make this configurable through a YAML config file
        c['data_backend']  = ENV['VISAGE_DATA_BACKEND'] || 'RRD'

        # RRD specific config options
        # FIXME: make this configurable through a YAML config file
        c['rrddir']        = ENV["RRDDIR"] ? Pathname.new(ENV["RRDDIR"]).expand_path : Pathname.new("/var/lib/collectd/rrd").expand_path
        c['collectdsock']  = ENV["COLLECTDSOCK"]
        c['rrdcachedsock'] = ENV["RRDCACHEDSOCK"]
      end

      # Upgrade the profile if we're running an older version
      if Visage::Upgrade.pending?
        puts "The Visage profile storage format has changed."
        upgrades = Visage::Upgrade.run
        first    = upgrades.first.version - 1
        last     = upgrades.last.version
        puts "Upgraded profile storage format from version #{first} to #{last}"
      end

      # Set the data backend to use in Visage::JSON
      Visage::Data.backend = Visage::Config.data_backend
    end

    configure :development do
      register Sinatra::Reloader
    end

    configure :test do
      disable :logging
    end
  end

  class Profiles < Application
    get '/' do
      redirect '/profiles'
    end

    get '/profiles/new' do
      haml :profile
    end

    get '/profiles/share/:id' do
      @profile = Profile.get(params[:id])
      raise Sinatra::NotFound unless @profile

      haml :share, :layout => false
    end

    # Viewing a single profile
    get %r{/profiles/([^/\.]+).?([^/]+)?} do
      id     = params[:captures][0]
      format = params[:captures][1]

      @profile = Profile.get(id)
      raise Sinatra::NotFound unless @profile

      if format == 'json'
        @profile.to_json
      else
        haml :profile
      end
    end

    # Viewing all profiles
    get %r{/profiles/*} do
      named_options = {
        :anonymous => false,
        :sort      => :name,
      }
      @profiles  = Profile.all(named_options)

      anonymous_options = {
        :anonymous => true,
        :sort      => :created_at,
        :order     => :descending,
      }
      @anonymous = Profile.all(anonymous_options)

      haml :profiles
    end

    # Creating a new profile
    post '/profiles' do
      attributes = ::JSON.parse(request.body.read).symbolize_keys
      @profile = Profile.new(attributes)

      if @profile.save
        {'status' => 'ok', 'id' => @profile.id}.to_json
      else
        status 400 # Bad Request
        {'status' => 'error', 'errors' => @profile.errors}.to_json
      end
    end

    # Updating an existing profile
    post %r{/profiles/([^/\.]+).?([^/]+)?} do
      id     = params[:captures][0]
      format = params[:captures][1]

      @profile = Profile.get(id)
      raise Sinatra::NotFound unless @profile

      p params
      if format == 'json'
        attrs = ::JSON.parse(request.body.read).symbolize_keys
      else
        attrs = params['profile'].symbolize_keys
      end

      if @profile.update_attributes(attrs)
        {'status' => 'ok', 'id' => @profile.id}.to_json
      else
        status 400 # Bad Request
        {'status' => 'error', 'errors' => @profile.errors}.to_json
      end
    end
  end


  class Builder < Application
  end

  class JSON < Application

    # JSON data backend
    mime_type :json,  "application/json"
    mime_type :jsonp, "text/javascript"

    before do
      content_type :jsonp
    end

    # /data/:host/:plugin/:optional_plugin_instance
    get %r{/data/([^/]+)/([^/]+)((/[^/]+)*)} do
      content_type :json if headers["Content-Type"] =~ /text/

      host        = params[:captures][0].gsub("\0", "")
      plugin      = params[:captures][1].gsub("\0", "")
      instances   = params[:captures][2].gsub("\0", "")
      start       = params[:start]
      finish      = params[:finish]
      percentiles = params[:percentiles] ||= "false"
      resolution  = params[:resolution]

      options = {
        :rrddir        => Visage::Config.rrddir,
        :collectdsock  => Visage::Config.collectdsock,
        :rrdcachedsock => Visage::Config.rrdcachedsock
      }
      data = Visage::Data.new(options)

      query = {
        :host        => host,
        :plugin      => plugin,
        :instances   => instances,
        :start       => start,
        :finish      => finish,
        :percentiles => percentiles,
        :resolution  => resolution,
      }
      json = data.json(query)

      # If the request is cross-domain, we need to serve JSON-P.
      maybe_wrap_with_callback(json)
    end

    get %r{/data/([^/]+)} do
      content_type :json if headers["Content-Type"] =~ /text/

      query = {
        :hosts => params[:captures][0].gsub("\0", "")
      }
      metrics = Visage::Data.metrics(query)
      json = { :metrics => metrics }.to_json

      maybe_wrap_with_callback(json)
    end

    get %r{/data(/)*} do
      content_type :json if headers["Content-Type"] =~ /text/

      hosts = Visage::Data.hosts
      json  = { :hosts => hosts }.to_json

      maybe_wrap_with_callback(json)
    end

    # Wraps json with a callback method that JSON-P clients can call.
    def maybe_wrap_with_callback(json)
      params[:callback] ? params[:callback] + '(' + json + ')' : json
    end

  end
end
